### PR TITLE
sound localization

### PR DESCRIPTION
**Pull request by yantrabuddhi**
initial improved detection, may need parameter tweaks

### CODE REVIEW

File: face_track/face_track.py [hunk 1]
```diff
@@ -144,8 +147,20 @@ def __init__(self):
 
 		# Which face to look at
 		# rospy.Subscriber(self.TOPIC_FACE_TARGET, xxxFaceEvent, xxxself.face_event_cb)
-		rospy.Subscriber("perceived_text", String,
-			self.user_said_cb)
+		#rospy.Subscriber("perceived_text", String,
+		#	self.user_said_cb)
```
Comment (linas) [inline face_track/face_track.py]: can you plese stop commenting out code!  If the code is dead, then please remove it, instead of commenting it out.

If you are not done with development, please don't submit a pull request.

File: face_track/face_track.py [hunk 2]
```diff
@@ -144,8 +147,20 @@ def __init__(self):
...
+		self.speech_start_at=datetime.datetime.now()
+		self.speech_end_at=datetime.datetime.now()
```
Comment (linas) [inline face_track/face_track.py]: this cannot be right -- start and end occurring at the same time?

File: face_track/face_track.py [hunk 3]
```diff
@@ -19,12 +19,15 @@
...
 
+import datetime
+
 import rospy
 import tf
 import math
 import logging
 
 from std_msgs.msg import Int32, String
+from chatbot.msg import ChatMessage
```
Comment (linas) [inline face_track/face_track.py:28]: please stop importing ChatMessage.   This has been an issue that has been biting us for more tha  a year. I keep getting rid of ChatMessage, it keeps coming back to life. Can we PLEASE kill it forever-more?

This breaks the open-source version of the robot.

File: face_track/face_track.py [hunk 4]
```diff
@@ -144,8 +147,20 @@ def __init__(self):
...
+		self.speech_end_at=datetime.datetime.now()
+		self.speech_elapse_ms=0.0
+		self.speech_end_elapse_ms=0.0;
+		rospy.Subscriber("chatbot_speech", ChatMessage,
+			self.stt_cb)
+
+		# rospy.Subscriber("chatbot_speech", ChatMessage,
+		# 	self.user_said_cb)
+		# rospy.Subscriber("chat_events", String,
+		# 	self.speech_event_cb)
```
Comment (linas) [inline face_track/face_track.py]: please do not submit code that is commented out.  Either the code works, or it doesn't.  Comment lines should be for comments, and not for code.

File: face_track/face_track.py [hunk 5]
```diff
@@ -392,8 +407,26 @@ def track_face(self, faceid):
...
+		if msg.confidence >= 50:
+			self.atomo.who_said(msg.utterance)
+
+	# def user_said_cb(self,msg):
+	# 	if msg.confidence >= 50:
+	# 		t1 = (datetime.datetime.now() - self.speech_start_at)
+	# 		t2 = (datetime.datetime.now() - self.speech_end_at)
+	# 		self.speech_elapse_ms =  t1.total_seconds()* 1000
+	# 		self.speech_end_elapse_ms = t2.total_seconds()*1000
+	# 		self.atomo.who_spoke(msg.utterance.lower(),
```
Comment (linas) [inline face_track/face_track.py]: More commented out code.

Also, what does this have to do with face-tracking?   Perhaps a different py file should be created for this speech stuff.  Integration of speech and visual perception should not be happening in the python layers, anyway.

File: src/ros_commo.py [hunk 6]
```diff
@@ -287,8 +287,8 @@ def say_text(self, text_to_say):
 	# Unit test by saying
 	#   rostopic pub --once perceived_text std_msgs/String "Look afraid!"
 	#
-	def language_perceived_text_cb(self, text_heard):
-		self.puta.perceived_text(text_heard.data)
+	#def language_perceived_text_cb(self, text_heard):
+	#	self.puta.perceived_text(text_heard.data)
```
Comment (linas) [inline src/ros_commo.py:291]: more commented out code.

File: src/ros_commo.py [hunk 7]
```diff
@@ -402,8 +402,9 @@ def openpsi_control_cb(self, data):
 
 			self.puta.evaluate_scm(scm_str)
 
-	def pub_snd_face_id(faceid):
-		self.face_sound_pub.publish(faceid)
+	#not needed anymore
+	#def pub_snd_face_id(faceid):
+	#	self.face_sound_pub.publish(faceid)
```
Comment (linas) [inline src/ros_commo.py:408]: more commented out code.

File: src/ros_commo.py [hunk 8]
```diff
@@ -451,8 +452,9 @@ def __init__(self):
 		self.gaze_at_pub = rospy.Publisher("/opencog/gaze_at",
 			Int32, queue_size=1)
 
-		self.face_sound_pub = rospy.Publisher("/manyyears/face_id",
-			Int32, queue_size=1)
+		#method shifted
+		#self.face_sound_pub = rospy.Publisher("/manyyears/face_id",
+		#	Int32, queue_size=1)
```
Comment (linas) [inline src/ros_commo.py:457]: more commented out code.

File: src/ros_commo.py [hunk 9]
```diff
@@ -477,8 +479,10 @@ def __init__(self):
 		                                  String, queue_size=1)
 
 		# String text of what the robot heard (from TTS)
-		rospy.Subscriber("perceived_text", String,
-			self.language_perceived_text_cb)
+		#no longer needed as chatbot_speech is Used
+		#chatbot ai.py no longer usd so not published
+		#rospy.Subscriber("perceived_text", String,
+		#	self.language_perceived_text_cb)
```
Comment (linas) [inline src/ros_commo.py:485]: more commented out code.

File: src/time-map.scm [hunk 10]
```diff
@@ -113,12 +126,32 @@
 		(angle ax ay az bx by bz)
 	)
 )
+
+(define (angle_bw_sf xyz-a xx yy zz)
+	(let* ((sta (string-split xyz-a #\ ))
+		(ax (string->number(car sta)))
+		(ay (string->number(cadr sta)))
+		(az (string->number(caddr sta))))
```
Comment (linas) [inline src/time-map.scm]: once again, we seem to have a cascade of string-to-number and number-to-string conversions, for the same values, over and over.   Please make up your mind as to which you want to use. please stop converting all the time.

You added a number API to the core atomspace bindings, despite my rather strongly recommending that it was a bad idea, and would only lead to chaos.   Well, the above three lines of code are a proof of that chaos.   You've got to pick an API, and stick to it.

File: src/time-map.scm [hunk 11]
```diff
@@ -184,17 +218,81 @@
...
+	(GetLink (StateLink (ConceptNode "last person who spoke") (VariableNode "$fid")))
+	)
+)
+
+
+(define (snd-nearest-face xx yy zz)
+	(let* ((lst (get-visible-faces))
+			(falist (map
+				(lambda (x) (list
+					(string->number (cog-name x))
```
Comment (linas) [inline src/time-map.scm:151]: more sting to number conversion.

**Comment by linas** [comment]
I guess this code is OK, and is maybe mergeable; however, the large blocks of commented out code need to be cleaned up.   Also the repeated number-string conversions indicate some sort of design flaw.  I guess they don't need to be fixed right away, but they are a symptom of a mis-designed API somewhere.

**Comment by yantrabuddhi** [comment]
Hi Linas,
 I will remove all the comments soon. I am currently in state of
experimenting with lot of stuff and I also think things should be done
proper architecturally. I myself am not satisfied with the overall
architecture now, but due to time constraints and large number of tasks
inefficient methods sneak in, like string to number conversions(in some
cases I do not know how to avoid them, although with some re-design some
can be totally removed)

**Comment by yantrabuddhi** [comment]
Hi @linas ,
 I re-factored the code as you said - removed the comments + where possible removed string-number conversions + removed all extra code not being used

**Comment by yantrabuddhi** [comment]
@linas : also, I have to use chatmessage because perceived text is generated by listening to chatmessage from aiml python module, it is no longer their in new opencog chatbot and is no longer published
